src/controllers/hid/hidioglobaloutputreportfifo.cpp

```diff
@@ -0,0 +1,131 @@
+#include "controllers/hid/hidioglobaloutputreportfifo.h"
+
+#include <hidapi.h>
+
+#include "controllers/defs_controllers.h"
+#include "controllers/hid/legacyhidcontrollermappingfilehandler.h"
+#include "util/compatibility/qbytearray.h"
+#include "util/string.h"
+#include "util/time.h"
+#include "util/trace.h"
+
+namespace {
+constexpr int kReportIdSize = 1;
+constexpr int kMaxHidErrorMessageSize = 512;
+} // namespace
+
+HidIoGlobalOutputReportFifo::HidIoGlobalOutputReportFifo()
+        : m_indexOfLastSentReport(0),
+          m_indexOfLastCachedReport(0) {
+}
+
+void HidIoGlobalOutputReportFifo::addReportDatasetToFifo(const quint8 reportId,
+        const QByteArray& data,
+        const mixxx::hid::DeviceInfo& deviceInfo,
+        const RuntimeLoggingCategory& logOutput) {
+    auto cacheLock = lockMutex(&m_fifoMutex);
+
+    unsigned int indexOfReportToCache;
+
+    if (m_indexOfLastCachedReport + 1 < kSizeOfFifoInReports) {
+        indexOfReportToCache = m_indexOfLastCachedReport + 1;
+    } else {
+        indexOfReportToCache = 0;
+    }
+
+    // Handle the case, that the FIFO is full - which is an error case
+    if (m_indexOfLastSentReport == indexOfReportToCache) {
+        // If the FIFO is full, we skip the report dataset even
+        // in non-skipping mode, to keep the controller mapping thread
+        // responsive for InputReports from the controller.
+        // Alternative would be to block processing of the controller
+        // mapping thread, until the FIFO has space again.
+        qCWarning(logOutput)
+                << "FIFO overflow: Unable to add OutputReport " << reportId
+                << "to the global cache for non-skipping sending of OututReports for"
+                << deviceInfo.formatName();
+        return;
+    }
+
+    // First byte must always contain the correct ReportID-Byte,
+    // also after swapping
+    QByteArray cachedData;
+    cachedData.reserve(kReportIdSize + data.size());
+    cachedData.append(reportId);
+    cachedData.append(data);
+
+    // Swap report data to FIFO
+    cachedData.swap(m_outputReportFifo[indexOfReportToCache]);
```

Comment:
Here we allocate memory, swap the pointer to the memory and delete the old memory. 
Did you consider to reuse the memory?

Reply:
Here we've a FIFO that can store reports with different ID and therefore different length. In case of identical length of old and new report, I could reuse the memory of cause.